# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/MediaStreamWindowCapturer.cpp
# Commit: c54cb3c10992
# Full Hash: c54cb3c109922bc15fb6d0ca67e5f2a05980c1a5
# Author: Andreas Pehrson <apehrson@mozilla.com>
# Date: 2019-07-31 15:46:43
# Regressor Bug: 1493613
# File Overlap Count: 2
# Description:
#   Bug 1493613 - Move MediaStream control from DOMMediaStream to MediaStreamTrack. r=padenot
#   
#   This is inherently large, because modifying these bits of DOMMediaStream and
#   MediaStreamTrack affects all consumers and producers of all DOMMediaStreams and
#   MediaStreamTracks.
# ==============================================================================

diff -r 0b03dd9d20ac -r c54cb3c10992 dom/media/MediaStreamWindowCapturer.cpp
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/dom/media/MediaStreamWindowCapturer.cpp	Wed Jul 31 07:58:17 2019 +0000
@@ -0,0 +1,73 @@
+/* -*- Mode: C++; tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2 -*-*/
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "MediaStreamWindowCapturer.h"
+
+#include "AudioStreamTrack.h"
+#include "DOMMediaStream.h"
+#include "MediaStreamGraph.h"
+
+namespace mozilla {
+using dom::AudioStreamTrack;
+using dom::MediaStreamTrack;
+
+MediaStreamWindowCapturer::CapturedTrack::CapturedTrack(
+    MediaStreamTrack* aTrack, uint64_t aWindowID)
+    : mTrack(aTrack),
+      mPort(aTrack->Graph()->ConnectToCaptureStream(aWindowID,
+                                                    aTrack->GetStream())) {}
+
+MediaStreamWindowCapturer::CapturedTrack::~CapturedTrack() { mPort->Destroy(); }
+
+MediaStreamWindowCapturer::MediaStreamWindowCapturer(DOMMediaStream* aStream,
+                                                     uint64_t aWindowId)
+    : mStream(aStream), mWindowId(aWindowId) {
+  mStream->RegisterTrackListener(this);
+  nsTArray<RefPtr<AudioStreamTrack>> tracks;
+  mStream->GetAudioTracks(tracks);
+  for (const auto& t : tracks) {
+    if (t->Ended()) {
+      continue;
+    }
+    AddTrack(t);
+  }
+}
+
+MediaStreamWindowCapturer::~MediaStreamWindowCapturer() {
+  if (mStream) {
+    mStream->UnregisterTrackListener(this);
+  }
+}
+
+void MediaStreamWindowCapturer::NotifyTrackAdded(
+    const RefPtr<MediaStreamTrack>& aTrack) {
+  if (AudioStreamTrack* at = aTrack->AsAudioStreamTrack()) {
+    AddTrack(at);
+  }
+}
+
+void MediaStreamWindowCapturer::NotifyTrackRemoved(
+    const RefPtr<MediaStreamTrack>& aTrack) {
+  if (AudioStreamTrack* at = aTrack->AsAudioStreamTrack()) {
+    RemoveTrack(at);
+  }
+}
+
+void MediaStreamWindowCapturer::AddTrack(AudioStreamTrack* aTrack) {
+  if (aTrack->Ended()) {
+    return;
+  }
+  mTracks.AppendElement(MakeUnique<CapturedTrack>(aTrack, mWindowId));
+}
+
+void MediaStreamWindowCapturer::RemoveTrack(AudioStreamTrack* aTrack) {
+  for (size_t i = mTracks.Length(); i > 0; --i) {
+    if (mTracks[i - 1]->mTrack == aTrack) {
+      mTracks.RemoveElementAt(i - 1);
+      break;
+    }
+  }
+}
+}  // namespace mozilla